python/pyspark/sql/session.py

```diff
@@ -214,8 +214,12 @@ def __init__(self, sparkContext, jsparkSession=None):
         self._wrapped = SQLContext(self._sc, self, self._jwrapped)
         _monkey_patch_RDD(self)
         install_exception_handler()
-        if SparkSession._instantiatedContext is None:
-            SparkSession._instantiatedContext = self
+        # If we had an instantiated SparkSession attached with a SparkContext
+        # which is stopped now, we need to renew the instantiated SparkSession.
+        # Otherwise, we will use invalid SparkSession when we call Builder.getOrCreate.
+        if SparkSession._instantiatedSession is None \
+                or SparkSession._instantiatedSession._sc._jsc is None:
```

Comment:
Well, as I tried, this would case test failed.

`__init__` and `getOrCreate` are both legal paths. Once one calls `__init__` to get a `SparkSession`  and other calls `getOrCreate` later, the second one will be the different session than the first one, if the logic to update `_instantiatedSession` is different.